src/logic/girard.lean

```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2021 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import data.set.basic
+
+/-!
+# Girard's paradox
+
+Girard's paradox is a proof that `Type : Type` entails a contradiction. We can't say this directly
+in lean because `Type : Type 1` and it's not possible to give `Type` a different type via an axiom,
+so instead we axiomatize the behavior of the Pi type and application if the typing rule for Pi was
+`(Type → Type) → Type` instead of `(Type → Type) → Type 1`.
+
+Furthermore, we don't actually want false axioms in mathlib, so rather than introducing the axioms
+using `axiom` or `constant` declarations, we declare a `girard` class, which bundles up the
+hypotheses in Girard's paradox, such that an instance of `girard.{u]` asserts that universe `u`
+satisfies the assumptions, i.e. `Type u : Type u`, give or take. Then the paradox amounts to a
+proof that the `girard` class is empty.
+
+Based on Watkins' LF implementation of Hurkens' simplification of Girard's paradox:
+<http://www.cs.cmu.edu/~kw/research/hurkens95tlca.elf>.
+
+## Main statements
+
+* `girard.paradox : girard → false`: there are no Girard universes.
+-/
+
+universe u
+
+/-- A "Girard universe", which is a universe `u` such that `Pi : (Type u → Type u) → Type u`.
+Since we can't actually change the type of lean's `Π` operator, we define `pi`, `lam`, `app` and
+the `beta` rule instead. -/
+class girard :=
+(pi : (Type u → Type u) → Type u)
+(lam {A : Type u → Type u} : (∀ x, A x) → pi A)
+(app {A} : pi A → ∀ x, A x)
+(beta {A : Type u → Type u} (f : ∀ x, A x) (x) : app (lam f) x = f x)
+
+namespace girard
+variable [girard.{u}]
```

Comment:
It might be nice to also add that this assumption is given by `small.{u} (Type u)` from `logic.small`:
```
lemma cast_helper {A : Type u → Type u} (x y : Type u)
  (f : Π (x : Type u), A x)
  (yx : y = x) :
  cast (congr_arg A yx) (f y) = f x :=
begin
  subst yx,
  apply cast_eq
end

noncomputable def girard_from_small_type [small.{u} (Type u)] : girard.{u} :=
{ pi := λ f, Π (y : shrink (Type u)), f ((equiv_shrink (Type u)).symm y),
  lam := λ A p t, p _,
  app := λ A p t,
    eq.mpr (congr_arg A ((equiv_shrink (Type u)).left_inv t).symm) (p (equiv_shrink (Type u) t)),
  beta :=
  begin
    intros A f x,
    change cast _ (f _) = _,
    apply cast_helper,
    simp
  end }
```
(presumably `cast_helper` is in mathlib in some form but I couldn't find it!)